File: longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> last(256, -1);
        int head = 0, longSub = 0;
        int len = s.length();
        for(int i = 0; i < len; ++i)
        {
             if(last[s[i]] < head)
             {
                 longSub = max(longSub, i - head + 1);
             }
             else
             {
                 head = last[s[i]] + 1;
             }
             last[s[i]] = i;
        }
        
        return longSub;
    }
};

File: longest-substring-without-repeating-characters.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    ans, head = 0, 0
    map = Hash.new()
    for i in 0..s.length-1
        if !map[s[i]] || map[s[i]] < head
            ans = [ans, i - head + 1].max
        else
            head = map[s[i]] + 1
        end
        
        map[s[i]] = i
    end
    
    return ans
end